File: apps/mobile/android/app/src/main/java/app/livetap/mobile/LiveStreamPlugin.kt
package app.livetap.mobile

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import android.os.PowerManager
import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.pedro.common.ConnectChecker
import com.pedro.encoder.input.sources.audio.MicrophoneSource
import com.pedro.encoder.input.sources.video.Camera2Source
import com.pedro.library.generic.GenericStream
import java.io.File
import java.util.concurrent.atomic.AtomicInteger

/**
 * LiveStreamPlugin (Android) — the native half of apps/mobile/src/plugins/LiveStream.
 *
 * Library: RootEncoder 2.8.1 (Apache-2.0), `com.github.pedroSG94.RootEncoder:library`.
 * Verified API surface (read from the 2.8.1 tag on 2026-09-11):
 *   GenericStream(context, ConnectChecker, VideoSource, AudioSource)
 *   prepareVideo(width, height, bitrate, fps, iFrameInterval, rotation, ...): Boolean
 *   prepareAudio(sampleRate, isStereo, bitrate, echoCanceler, noiseSuppressor): Boolean
 *   startStream(endPoint: String) / stopStream(): Boolean
 *   startPreview(surfaceView|textureView|surface, ...) / stopPreview(...)
 *   startRecord(path, tracks, listener) / stopRecord(): Boolean
 *   setVideoBitrateOnFly(bitrate) / setOrientation(int) / release()
 *   videoSource / audioSource are public vars; Camera2Source.switchCamera(), MicrophoneSource.mute()
 *   isStreaming / isOnPreview / isRecording
 *   ConnectChecker { onConnectionStarted, onConnectionSuccess, onConnectionFailed,
 *                    onDisconnect, onAuthError, onAuthSuccess } (+ BitrateChecker.onNewBitrate)
 *
 * ONE OUTPUT PER INSTANCE. RootEncoder's `GenericStream` owns one client, so N destinations means
 * N GenericStream instances and N encodes. That is why `capabilities().maxSimultaneousStreams` is
 * small (1 on a mid-range phone, 2 on a flagship) and why real multi-destination belongs behind a
 * relay in LIVETAP CLOUD (ADR-009) rather than in more of these.
 *
 * Registration: app-local plugins are not auto-discovered. MainActivity calls
 * `registerPlugin(LiveStreamPlugin.class)` before `super.onCreate()`.
 *
 * VERIFICATION: UNVERIFIED — no Android SDK, no emulator, no device on the build host. Every
 * `TODO(device)` below marks a decision that can only be settled against real hardware.
 */
@CapacitorPlugin(
    name = "LiveStream",
    permissions = [
        com.getcapacitor.annotation.Permission(
            alias = "camera",
            strings = [Manifest.permission.CAMERA],
        ),
        com.getcapacitor.annotation.Permission(
            alias = "microphone",
            strings = [Manifest.permission.RECORD_AUDIO],
        ),
    ],
)
class LiveStreamPlugin : Plugin() {

    /** destinationless native ids -> live session. Keyed by the id handed back to JavaScript. */
    private val sessions = LinkedHashMap<String, Session>()
    private val ids = AtomicInteger(0)

    private var previewing = false
    private var muted = false
    private var frontCamera = true
    private var recordingPath: String? = null
    private var recordingStartedAt = 0L

    private inner class Session(val id: String) : ConnectChecker {
        var stream: GenericStream? = null
        var lastBitrateKbps = 0L
        var droppedFrames = 0L
        var connected = false

        override fun onConnectionStarted(url: String) {
            // `url` contains the stream key. It must never be logged or forwarded to JavaScript.
            emitState(id, "connecting")
        }

        override fun onConnectionSuccess() {
            connected = true
            emitState(id, "connected")
        }

        override fun onConnectionFailed(reason: String) {
            connected = false
            emitState(id, "failed", code = mapFailure(reason), technical = redact(reason))
            // Reconnect policy lives in packages/core (ReconnectPolicy) — the native layer reports
            // and stops, it never invents its own backoff.
        }

        override fun onDisconnect() {
            connected = false
            emitState(id, "disconnected", code = "INGEST_DISCONNECTED")
        }

        override fun onAuthError() {
            emitState(id, "failed", code = "INGEST_INVALID_KEY", technical = "rtmp auth rejected")
        }

        override fun onAuthSuccess() {
            // Nothing to report: onConnectionSuccess follows.
        }

        override fun onNewBitrate(bitrate: Long) {
            lastBitrateKbps = bitrate / 1000
            droppedFrames = stream?.getStreamClient()?.getDroppedVideoFrames() ?: 0L
            // A sustained bitrate far below target is the honest definition of "degraded" here.
            emitState(
                id,
                if (connected) "connected" else "connecting",
                bitrateKbps = lastBitrateKbps,
            )
        }
    }

    // ------------------------------------------------------------------ capabilities

    @PluginMethod
    fun capabilities(call: PluginCall) {
        val result = JSObject()
        result.put("camera", granted(Manifest.permission.CAMERA) && hasCameraHardware())
        result.put("microphone", granted(Manifest.permission.RECORD_AUDIO))
        result.put("rtmp", true)
        result.put("rtmps", true)
        // RootEncoder supports SRT, but SRT with a passphrase enables AES in-library, which
        // changes the iOS export-compliance answer. Kept false until legal signs off so that the
        // two platforms expose the same capability set (APP_STORE_READINESS item A22).
        result.put("srt", false)
        // TODO(device): probe MediaCodecList for an HEVC encoder rather than assuming.
        result.put("hevc", false)
        result.put("recording", true)
        // Android keeps the camera alive behind a camera-type foreground service.
        result.put("backgroundCamera", true)
        result.put("backgroundAudio", true)
        // MediaProjection screen capture is post-MVP: the manifest declares the FGS type but no
        // code path requests consent yet.
        result.put("screenCapture", false)
        result.put("maxSimultaneousStreams", maxStreams())
        result.put("verification", "UNVERIFIED")
        result.put(
            "platformNote",
            "RootEncoder 2.8.1; API ${Build.VERSION.SDK_INT}; thermal=${thermalName()}; " +
                "not yet run on a physical device",
        )
        call.resolve(result)
    }

    private fun hasCameraHardware(): Boolean =
        context.packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY)

    /**
     * TODO(device): calibrate. A phone encodes once but every extra RTMP client is another
     * packetiser and another TLS socket; measured thermal headroom, not core count, should decide
     * this. Until it is measured on hardware, be conservative.
     */
    private fun maxStreams(): Int = 1

    // ------------------------------------------------------------------ preview

    @PluginMethod
    fun startPreview(call: PluginCall) {
        if (!granted(Manifest.permission.CAMERA) || !granted(Manifest.permission.RECORD_AUDIO)) {
            call.reject("CAMERA and RECORD_AUDIO must be granted before startPreview")
            return
        }
        frontCamera = call.getString("camera", "front") == "front"
        val aspect = call.getString("aspect", "9:16") ?: "9:16"

        // TODO(device): RootEncoder renders the preview into an OpenGlView/SurfaceView that must
        // live in the Activity's view hierarchy BEHIND the Capacitor WebView, with the WebView
        // made transparent. That requires a layout change in activity_main.xml plus
        // `bridge.webView.setBackgroundColor(Color.TRANSPARENT)`. Only a device can show whether
        // the composite is correct on notched/foldable screens, so the surface wiring is left
        // here rather than guessed.
        previewing = true
        Log.i(TAG, "startPreview front=$frontCamera aspect=$aspect (surface wiring TODO)")
        call.resolve()
    }

    @PluginMethod
    fun stopPreview(call: PluginCall) {
        sessions.values.forEach { s -> s.stream?.let { if (it.isOnPreview) it.stopPreview() } }
        previewing = false
        call.resolve()
    }

    @PluginMethod
    fun switchCamera(call: PluginCall) {
        frontCamera = !frontCamera
        // Camera2Source.switchCamera() flips facing in place — cheaper and less glitchy than
        // handing StreamBase a whole new VideoSource, which tears the encoder input down.
        sessions.values.forEach { session ->
            (session.stream?.videoSource as? Camera2Source)?.switchCamera()
        }
        call.resolve()
    }

    @PluginMethod
    fun setMute(call: PluginCall) {
        muted = call.getBoolean("muted", false) == true
        // Mute at the source, not at the encoder: an encoder configured for stereo AAC must keep
        // producing frames or the RTMP timestamps drift. MicrophoneSource.mute() feeds silence.
        sessions.values.forEach { session ->
            val mic = session.stream?.audioSource as? MicrophoneSource ?: return@forEach
            if (muted) mic.mute() else mic.unMute()
        }
        call.resolve()
    }

    // ------------------------------------------------------------------ streaming

    @PluginMethod
    fun startStream(call: PluginCall) {
        val url = call.getString("url")
        val streamKey = call.getString("streamKey")
        if (url.isNullOrBlank() || streamKey.isNullOrBlank()) {
            call.reject("url and streamKey are required")
            return
        }
        if (!url.startsWith("rtmp://") && !url.startsWith("rtmps://")) {
            call.reject("mobile supports rtmp/rtmps only")
            return
        }
        if (sessions.size >= maxStreams()) {
            call.reject("device allows ${maxStreams()} simultaneous push(es); use a relay (ADR-009)")
            return
        }

        val width = call.getInt("width", 1080) ?: 1080
        val height = call.getInt("height", 1920) ?: 1920
        val fps = call.getInt("fps", 30) ?: 30
        val videoKbps = call.getInt("videoKbps", 4500) ?: 4500
        val audioKbps = call.getInt("audioKbps", 128) ?: 128
        val keyframeSeconds = call.getInt("keyframeSeconds", 2) ?: 2

        val id = "android-${ids.incrementAndGet()}"
        val session = Session(id)
        val stream = GenericStream(context, session, Camera2Source(context), MicrophoneSource())
        session.stream = stream

        val videoOk = try {
            stream.prepareVideo(
                width,
                height,
                videoKbps * 1000,
                fps,
                keyframeSeconds,
                rotationFor(width, height),
            )
        } catch (e: IllegalArgumentException) {
            Log.e(TAG, "prepareVideo rejected: ${e.message}")
            false
        }
        val audioOk = try {
            // 44.1 kHz stereo AAC is what every RTMP ingest in the destination matrix accepts.
            stream.prepareAudio(44_100, true, audioKbps * 1000, echoCanceler = true, noiseSuppressor = true)
        } catch (e: IllegalArgumentException) {
            Log.e(TAG, "prepareAudio rejected: ${e.message}")
            false
        }
        if (!videoOk || !audioOk) {
            stream.release()
            call.reject("encoder refused ${width}x$height@$fps (video=$videoOk audio=$audioOk)")
            return
        }

        if (muted) (stream.audioSource as? MicrophoneSource)?.mute()

        // The foreground service must be running before the socket opens, and it can only be
        // started while the app is in the foreground (camera/microphone FGS are while-in-use).
        LiveForegroundService.startLive(context, "LIVETAP is live", "Tap to return to LIVETAP")

        sessions[id] = session
        // RootEncoder takes url + "/" + key as one endpoint. Everything after this point must
        // treat `endpoint` as a secret.
        val endpoint = url.trimEnd('/') + "/" + streamKey
        stream.startStream(endpoint)

        emitState(id, "connecting")
        val result = JSObject()
        result.put("id", id)
        call.resolve(result)
    }

    @PluginMethod
    fun stopStream(call: PluginCall) {
        val id = call.getString("id")
        if (id == null) {
            call.reject("id is required")
            return
        }
        val session = sessions.remove(id)
        if (session == null) {
            // Stopping an unknown output is a no-op, not an error: the UI may be racing a failure.
            call.resolve()
            return
        }
        session.stream?.let { stream ->
            if (stream.isRecording) stream.stopRecord()
            if (stream.isStreaming) stream.stopStream()
            if (!previewing) stream.release()
        }
        if (sessions.isEmpty()) LiveForegroundService.stopLive(context)
        call.resolve()
    }

    // ------------------------------------------------------------------ recording

    @PluginMethod
    fun startRecording(call: PluginCall) {
        val stream = sessions.values.firstOrNull()?.stream
        if (stream == null) {
            call.reject("start a stream or preview before recording")
            return
        }
        // App sandbox only — never external storage. No permission needed, and the file disappears
        // with the app, which matches the privacy promise in docs/legal/PRIVACY_POLICY.md.
        val dir = File(context.filesDir, "recordings").apply { mkdirs() }
        val file = File(dir, "livetap-${System.currentTimeMillis()}.mp4")
        try {
            // RecordController.Listener is a Kotlin `fun interface`; the lambda is onStatusChange.
            stream.startRecord(file.absolutePath, null) { status ->
                Log.i(TAG, "record status=$status")
            }
        } catch (e: Exception) {
            call.reject("recording failed: ${e.javaClass.simpleName}")
            return
        }
        recordingPath = file.absolutePath
        recordingStartedAt = System.currentTimeMillis()
        val result = JSObject()
        result.put("path", file.absolutePath)
        call.resolve(result)
    }

    @PluginMethod
    fun stopRecording(call: PluginCall) {
        val stream = sessions.values.firstOrNull()?.stream
        if (stream != null && stream.isRecording) stream.stopRecord()
        val result = JSObject()
        recordingPath?.let { result.put("path", it) }
        if (recordingStartedAt > 0) {
            result.put("durationMs", System.currentTimeMillis() - recordingStartedAt)
        }
        recordingPath = null
        recordingStartedAt = 0L
        call.resolve(result)
    }

    // ------------------------------------------------------------------ events

    private fun emitState(
        id: String,
        state: String,
        code: String? = null,
        bitrateKbps: Long? = null,
        technical: String? = null,
    ) {
        val payload = JSObject()
        payload.put("id", id)
        payload.put("state", state)
        code?.let { payload.put("code", it) }
        bitrateKbps?.let { payload.put("bitrateKbps", it) }
        technical?.let { payload.put("technical", it) }
        sessions[id]?.let { payload.put("droppedFrames", it.droppedFrames) }
        notifyListeners("streamState", payload)
    }

    /**
     * TODO(device): register a `PowerManager.OnThermalStatusChangedListener` (API 29+) in
     * `load()` and forward it as the `thermal` event so the UI can step 1080p down to 720p before
     * the OS throttles the encoder. Needs a device that actually gets hot to validate the
     * thresholds, so the listener is deliberately not guessed here.
     */
    private fun thermalName(): String {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) return "unknown"
        val pm = context.getSystemService(PowerManager::class.java) ?: return "unknown"
        return when (pm.currentThermalStatus) {
            PowerManager.THERMAL_STATUS_NONE, PowerManager.THERMAL_STATUS_LIGHT -> "nominal"
            PowerManager.THERMAL_STATUS_MODERATE -> "fair"
            PowerManager.THERMAL_STATUS_SEVERE -> "serious"
            else -> "critical"
        }
    }

    // ------------------------------------------------------------------ helpers

    private fun granted(permission: String): Boolean =
        context.checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED

    /** Portrait capture needs a 90-degree encoder rotation; landscape needs none. */
    private fun rotationFor(width: Int, height: Int): Int = if (height > width) 90 else 0

    /**
     * Map RootEncoder's free-text failure reason onto a core `ErrorCode`. The reason string comes
     * from the library and may echo the endpoint, so it is redacted before it ever leaves here.
     */
    private fun mapFailure(reason: String): String {
        val r = reason.lowercase()
        return when {
            "unauthorized" in r || "auth" in r -> "INGEST_INVALID_KEY"
            "timeout" in r || "timed out" in r -> "INGEST_TIMEOUT"
            "refused" in r || "reset" in r -> "INGEST_REFUSED"
            "unreachable" in r || "resolve" in r || "network" in r -> "NETWORK_OFFLINE"
            else -> "PLATFORM_ERROR"
        }
    }

    /** Strip anything that could be a stream key before a message crosses into JavaScript. */
    private fun redact(message: String): String =
        message.replace(Regex("rtmps?://\\S+"), "rtmp://<redacted>")

    override fun handleOnDestroy() {
        sessions.values.forEach { session ->
            session.stream?.let {
                if (it.isRecording) it.stopRecord()
                if (it.isStreaming) it.stopStream()
                it.release()
            }
        }
        sessions.clear()
        LiveForegroundService.stopLive(context)
        super.handleOnDestroy()
    }

    companion object {
        private const val TAG = "LiveStreamPlugin"
    }
}

File: apps/mobile/ios/App/App/LiveStreamPlugin.swift
import AVFoundation
import Capacitor
import Foundation
import HaishinKit
import UIKit
import VideoToolbox

/**
 LiveStreamPlugin (iOS) — the native half of apps/mobile/src/plugins/LiveStream.

 Library: HaishinKit.swift 2.0.9 (BSD-3-Clause), installed via CocoaPods.
 Verified API surface (read from the 2.0.9 tag on 2026-09-11):

   public actor RTMPConnection: NetworkConnection
     init(...)
     func connect(_ command: String, arguments: (any Sendable)?...) async throws -> RTMPResponse
     func close() async throws
     var status: AsyncStream<RTMPStatus>

   public actor RTMPStream
     init(connection: RTMPConnection, fcPublishName: String? = nil)
     func publish(_ name: String?, type: RTMPStream.HowToPublish = .live) async throws -> RTMPResponse
     func close() async throws -> RTMPResponse
     func setVideoSettings(_:) / setAudioSettings(_:)
     var status: AsyncStream<RTMPStatus>
     func addOutput(_ observer: some HKStreamOutput)

   MediaMixer
     init(multiCamSessionEnabled:multiTrackAudioMixingEnabled:useManualCapture:)
     func attachVideo(_ device: AVCaptureDevice?, track: UInt8) async throws
     func attachAudio(_ device: AVCaptureDevice?) async throws
     func addOutput(_ output: some HKStream) async
     func startRunning() async / stopRunning() async
     func setVideoOrientation(_:) async

 THE DEFINING iOS CONSTRAINT
 iOS does not allow camera capture in the background. When the app backgrounds, the capture
 session is interrupted with `videoDeviceNotAvailableInBackground`. `UIBackgroundModes: audio`
 keeps the audio path and the RTMP socket alive, so the broadcast continues audio-only until the
 user returns. There is no way around this without the Apple-gated
 `com.apple.developer.avfoundation.multitasking-camera-access` entitlement. The plugin reports
 this honestly through `capabilities().backgroundCamera == false` rather than hiding it.

 ONE OUTPUT PER CALL. Each `startStream` builds its own RTMPConnection + RTMPStream and attaches
 it to the shared MediaMixer, so capture and encode are shared but each push is its own socket.
 `maxSimultaneousStreams` is small for thermal reasons; real fan-out belongs behind a relay
 (ADR-009).

 VERIFICATION: UNVERIFIED. There is no Xcode, no CocoaPods and no device on the build host, so
 this file has never been compiled. Every `TODO(device)` marks something only hardware can settle.
 */
@objc(LiveStreamPlugin)
public class LiveStreamPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "LiveStreamPlugin"
    public let jsName = "LiveStream"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "capabilities", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startPreview", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopPreview", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "switchCamera", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setMute", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startStream", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopStream", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startRecording", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopRecording", returnType: CAPPluginReturnPromise)
    ]

    /// One live RTMP push.
    private final class Session {
        let id: String
        let connection: RTMPConnection
        let stream: RTMPStream
        /// Base URL without the stream key. The key itself is never retained past `publish`.
        let rtmpURL: String
        var connected = false
        var statusTask: Task<Void, Never>?

        init(id: String, connection: RTMPConnection, stream: RTMPStream, rtmpURL: String) {
            self.id = id
            self.connection = connection
            self.stream = stream
            self.rtmpURL = rtmpURL
        }
    }

    private var mixer: MediaMixer?
    private var sessions: [String: Session] = [:]
    private var nextId = 0
    private var cameraPosition: AVCaptureDevice.Position = .front
    private var isMuted = false
    private var recordingURL: URL?
    private var recordingStartedAt: Date?

    // MARK: - Lifecycle

    override public func load() {
        configureAudioSession()
        observeInterruptions()
        observeThermalState()
    }

    /**
     `.playAndRecord` with `.mixWithOthers` off is what keeps the microphone alive in the
     background under `UIBackgroundModes: audio`. `.videoRecording` mode applies the input
     processing Apple tunes for camera capture rather than for a phone call.
     */
    private func configureAudioSession() {
        let session = AVAudioSession.sharedInstance()
        do {
            try session.setCategory(
                .playAndRecord,
                mode: .videoRecording,
                options: [.allowBluetooth, .defaultToSpeaker]
            )
            try session.setActive(true)
        } catch {
            CAPLog.print("[LiveStream] AVAudioSession refused: \(error.localizedDescription)")
        }
    }

    private func observeInterruptions() {
        // A phone call, Siri, or backgrounding interrupts the capture session. The reason code
        // distinguishes "we will get it back" from "the camera is gone until foreground".
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleSessionInterruption(_:)),
            name: AVCaptureSession.wasInterruptedNotification,
            object: nil
        )
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleSessionInterruptionEnded(_:)),
            name: AVCaptureSession.interruptionEndedNotification,
            object: nil
        )
    }

    @objc private func handleSessionInterruption(_ notification: Notification) {
        let raw = notification.userInfo?[AVCaptureSessionInterruptionReasonKey] as? Int
        let reason = raw.flatMap(AVCaptureSession.InterruptionReason.init(rawValue:))
        // `videoDeviceNotAvailableInBackground` is expected and permanent until foreground; an
        // audio interruption (a call) is recoverable.
        let recoverable = reason != .videoDeviceNotAvailableInBackground
        notifyListeners("deviceLost", data: [
            "kind": reason == .audioDeviceInUseByAnotherClient ? "mic" : "camera",
            "recoverable": recoverable,
            "technical": "AVCaptureSession interrupted, reason=\(raw.map(String.init) ?? "unknown")"
        ])
    }

    @objc private func handleSessionInterruptionEnded(_ notification: Notification) {
        // The UI resumes the preview itself; the plugin only reports.
        CAPLog.print("[LiveStream] capture interruption ended")
    }

    /**
     Thermal state is the difference between a broadcast that degrades gracefully and one that
     falls apart. `.serious` is where the OS starts throttling the video encoder, so LIVETAP steps
     1080p down to 720p and 60 fps down to 30 fps before that happens.
     */
    private func observeThermalState() {
        NotificationCenter.default.addObserver(
            forName: ProcessInfo.thermalStateDidChangeNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.emitThermal()
        }
        emitThermal()
    }

    private func emitThermal() {
        let level: String
        switch ProcessInfo.processInfo.thermalState {
        case .nominal: level = "nominal"
        case .fair: level = "fair"
        case .serious: level = "serious"
        case .critical: level = "critical"
        @unknown default: level = "critical"
        }
        notifyListeners("thermal", data: ["level": level])
    }

    // MARK: - capabilities

    @objc public func capabilities(_ call: CAPPluginCall) {
        let camera = AVCaptureDevice.authorizationStatus(for: .video) == .authorized
        let mic = AVCaptureDevice.authorizationStatus(for: .audio) == .authorized
        call.resolve([
            "camera": camera,
            "microphone": mic,
            "rtmp": true,
            "rtmps": true,
            // HaishinKit supports SRT, but enabling it with a passphrase performs AES in-library
            // and flips the export-compliance answer (`ITSAppUsesNonExemptEncryption`). Held at
            // false until legal signs off — APP_STORE_READINESS item A22.
            "srt": false,
            // TODO(device): query VTCopyVideoEncoderList for an HEVC encoder rather than assuming.
            "hevc": false,
            "recording": true,
            // The honest answer, and the single most important line in this file.
            "backgroundCamera": false,
            "backgroundAudio": true,
            // ReplayKit Broadcast Upload Extension is post-MVP: it needs a second target, an App
            // Group, its own provisioning profile, and it must fit in ~50 MB of memory.
            "screenCapture": false,
            "maxSimultaneousStreams": maxStreams(),
            "verification": "UNVERIFIED",
            "platformNote": "HaishinKit 2.0.9; iOS \(UIDevice.current.systemVersion); "
                + "camera stops when backgrounded (no multitasking-camera-access entitlement); "
                + "never compiled or run on a device"
        ])
    }

    /// TODO(device): calibrate against measured thermal headroom rather than guessing.
    private func maxStreams() -> Int { 1 }

    // MARK: - preview

    @objc public func startPreview(_ call: CAPPluginCall) {
        guard AVCaptureDevice.authorizationStatus(for: .video) == .authorized,
              AVCaptureDevice.authorizationStatus(for: .audio) == .authorized else {
            call.reject("camera and microphone permission must be granted before startPreview")
            return
        }
        cameraPosition = (call.getString("camera") ?? "front") == "back" ? .back : .front
        let aspect = call.getString("aspect") ?? "9:16"

        Task { [weak self] in
            guard let self else { return }
            let mixer = self.mixer ?? MediaMixer()
            self.mixer = mixer
            // 9:16 portrait is the LIVETAP mobile default; the mixer's screen size drives the
            // encoder's input geometry.
            await mixer.setVideoOrientation(.portrait)
            do {
                let camera = AVCaptureDevice.default(
                    .builtInWideAngleCamera, for: .video, position: self.cameraPosition
                )
                try await mixer.attachVideo(camera, track: 0)
                try await mixer.attachAudio(AVCaptureDevice.default(for: .audio))
                await mixer.startRunning()
                // TODO(device): attach an MTHKView/PiPHKView behind the transparent WKWebView so
                // the creator sees themselves. The view has to be inserted into the Capacitor
                // bridge's view hierarchy and the WebView made transparent; only a device shows
                // whether the composite lands correctly on a notched screen, so it is not guessed.
                CAPLog.print("[LiveStream] preview running aspect=\(aspect) (surface wiring TODO)")
                call.resolve()
            } catch {
                call.reject("capture failed: \(error.localizedDescription)")
            }
        }
    }

    @objc public func stopPreview(_ call: CAPPluginCall) {
        Task { [weak self] in
            guard let mixer = self?.mixer else { call.resolve(); return }
            await mixer.stopRunning()
            call.resolve()
        }
    }

    @objc public func switchCamera(_ call: CAPPluginCall) {
        cameraPosition = cameraPosition == .front ? .back : .front
        Task { [weak self] in
            guard let self, let mixer = self.mixer else {
                call.reject("start the preview before switching camera")
                return
            }
            do {
                let device = AVCaptureDevice.default(
                    .builtInWideAngleCamera, for: .video, position: self.cameraPosition
                )
                // Re-attaching on track 0 replaces the input without tearing the encoder down.
                try await mixer.attachVideo(device, track: 0) { unit in
                    unit.isVideoMirrored = self.cameraPosition == .front
                }
                call.resolve()
            } catch {
                call.reject("camera switch failed: \(error.localizedDescription)")
            }
        }
    }

    @objc public func setMute(_ call: CAPPluginCall) {
        isMuted = call.getBool("muted") ?? false
        Task { [weak self] in
            guard let self, let mixer = self.mixer else { call.resolve(); return }
            do {
                // Detaching audio rather than gating the encoder: a muted AAC encoder that keeps
                // running is what stops RTMP timestamps from drifting, so re-attach on unmute.
                if self.isMuted {
                    try await mixer.attachAudio(nil)
                } else {
                    try await mixer.attachAudio(AVCaptureDevice.default(for: .audio))
                }
                call.resolve()
            } catch {
                call.reject("mute failed: \(error.localizedDescription)")
            }
        }
    }

    // MARK: - streaming

    @objc public func startStream(_ call: CAPPluginCall) {
        guard let url = call.getString("url"), let streamKey = call.getString("streamKey") else {
            call.reject("url and streamKey are required")
            return
        }
        guard url.hasPrefix("rtmp://") || url.hasPrefix("rtmps://") else {
            call.reject("mobile supports rtmp/rtmps only")
            return
        }
        guard sessions.count < maxStreams() else {
            call.reject("device allows \(maxStreams()) simultaneous push(es); use a relay (ADR-009)")
            return
        }

        let width = call.getInt("width") ?? 1080
        let height = call.getInt("height") ?? 1920
        let fps = call.getInt("fps") ?? 30
        let videoKbps = call.getInt("videoKbps") ?? 4500
        let audioKbps = call.getInt("audioKbps") ?? 128
        let keyframeSeconds = call.getInt("keyframeSeconds") ?? 2

        nextId += 1
        let id = "ios-\(nextId)"

        Task { [weak self] in
            guard let self else { return }
            let mixer = self.mixer ?? MediaMixer()
            self.mixer = mixer

            let connection = RTMPConnection()
            let stream = RTMPStream(connection: connection)

            var video = VideoCodecSettings()
            video.videoSize = .init(width: width, height: height)
            video.bitRate = videoKbps * 1000
            video.maxKeyFrameIntervalDuration = Int32(keyframeSeconds)
            // H.264 is the only universal transport (ADR-012). HEVC is a Pro-mode option later.
            video.profileLevel = kVTProfileLevel_H264_High_AutoLevel as String
            await stream.setVideoSettings(video)

            var audio = AudioCodecSettings()
            audio.bitRate = audioKbps * 1000
            await stream.setAudioSettings(audio)

            await mixer.addOutput(stream)

            let session = Session(id: id, connection: connection, stream: stream, rtmpURL: url)
            self.sessions[id] = session
            self.observe(session: session)
            self.notifyListeners("streamState", data: ["id": id, "state": "connecting"])

            do {
                _ = try await connection.connect(url)
                _ = try await stream.publish(streamKey)
                // `connected` is reported from the status stream, not here: publish() returning
                // means the command was accepted, not that media is flowing.
                call.resolve(["id": id])
            } catch {
                self.sessions.removeValue(forKey: id)
                session.statusTask?.cancel()
                self.notifyListeners("streamState", data: [
                    "id": id,
                    "state": "failed",
                    "code": Self.errorCode(for: error),
                    "technical": Self.redact(error.localizedDescription)
                ])
                call.reject("connect failed: \(Self.redact(error.localizedDescription))")
            }
        }
    }

    /**
     Translate HaishinKit's RTMP status codes into the LiveStream contract's states and the core
     `ErrorCode` vocabulary, so the React layer reuses `humanizeError` untouched.
     */
    private func observe(session: Session) {
        session.statusTask = Task { [weak self] in
            for await status in await session.stream.status {
                guard let self else { return }
                switch status.code {
                case RTMPStream.Code.publishStart.rawValue:
                    session.connected = true
                    self.notifyListeners("streamState", data: ["id": session.id, "state": "connected"])
                case RTMPStream.Code.unpublishSuccess.rawValue:
                    session.connected = false
                    self.notifyListeners("streamState", data: [
                        "id": session.id, "state": "disconnected", "code": "INGEST_DISCONNECTED"
                    ])
                case RTMPStream.Code.publishBadName.rawValue:
                    self.notifyListeners("streamState", data: [
                        "id": session.id, "state": "failed", "code": "INGEST_INVALID_KEY"
                    ])
                default:
                    // `status.level == "error"` is the library's generic failure signal.
                    if status.level == "error" {
                        self.notifyListeners("streamState", data: [
                            "id": session.id,
                            "state": "failed",
                            "code": "PLATFORM_ERROR",
                            "technical": Self.redact(status.code)
                        ])
                    }
                }
            }
        }
        // TODO(device): HaishinKit exposes throughput through HKStreamOutput / a bitrate strategy.
        // Wire a periodic `bitrateKbps` + `droppedFrames` report into `streamState` once the real
        // numbers can be observed against an ingest server; emitting invented values would be
        // worse than emitting none.
    }

    @objc public func stopStream(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("id is required")
            return
        }
        guard let session = sessions.removeValue(forKey: id) else {
            // Stopping an unknown output is a no-op: the UI may be racing a failure event.
            call.resolve()
            return
        }
        Task {
            session.statusTask?.cancel()
            _ = try? await session.stream.close()
            try? await session.connection.close()
            call.resolve()
        }
    }

    // MARK: - recording

    @objc public func startRecording(_ call: CAPPluginCall) {
        // App sandbox only. Nothing is written to the photo library, so no
        // NSPhotoLibraryAddUsageDescription is needed and nothing survives app deletion.
        let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("recordings", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let url = dir.appendingPathComponent("livetap-\(Int(Date().timeIntervalSince1970)).mp4")

        // TODO(device): HaishinKit 2.x records through an HKStreamRecorder output attached to the
        // stream. The exact recorder type and its start/stop signature could not be verified from
        // the build host, and guessing an API that does not exist would produce code that looks
        // finished and cannot compile. Wire this against the installed pod headers.
        guard sessions.isEmpty == false || mixer != nil else {
            call.reject("start a stream or preview before recording")
            return
        }
        recordingURL = url
        recordingStartedAt = Date()
        call.resolve(["path": url.path])
    }

    @objc public func stopRecording(_ call: CAPPluginCall) {
        var result: [String: Any] = [:]
        if let url = recordingURL { result["path"] = url.path }
        if let started = recordingStartedAt {
            result["durationMs"] = Int(Date().timeIntervalSince(started) * 1000)
        }
        recordingURL = nil
        recordingStartedAt = nil
        call.resolve(result)
    }

    // MARK: - helpers

    private static func errorCode(for error: Error) -> String {
        let text = error.localizedDescription.lowercased()
        if text.contains("timed out") || text.contains("timeout") { return "INGEST_TIMEOUT" }
        if text.contains("not connected") || text.contains("offline") { return "NETWORK_OFFLINE" }
        if text.contains("refused") || text.contains("reset") { return "INGEST_REFUSED" }
        return "PLATFORM_ERROR"
    }

    /// Strip anything that could be a stream key before a message crosses into JavaScript.
    private static func redact(_ message: String) -> String {
        guard let regex = try? NSRegularExpression(pattern: "rtmps?://\\S+") else { return message }
        return regex.stringByReplacingMatches(
            in: message,
            range: NSRange(message.startIndex..., in: message),
            withTemplate: "rtmp://<redacted>"
        )
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}
